File: modules/wem-api/src/main/java/com/enonic/wem/api/data/PropertyVisitor.java
package com.enonic.wem.api.data;


public abstract class PropertyVisitor
{
    private ValueType valueType;

    private boolean visitPropertiesWithSet = false;

    public PropertyVisitor restrictType( final ValueType valueType )
    {
        this.valueType = valueType;
        return this;
    }

    public PropertyVisitor visitPropertiesWithSet( final boolean value )
    {
        visitPropertiesWithSet = value;
        return this;
    }

    public void traverse( final PropertyTree propertyTree )
    {
        traverse( propertyTree.getRoot() );
    }

    public void traverse( final PropertySet propertySet )
    {
        for ( final Property property : propertySet.getProperties() )
        {
            final boolean valueTypeRestrictionSatisfied = valueType == null || valueType.equals( property.getValue().getType() );
            final boolean propertyWithSetRestrictionSatisfied =
                visitPropertiesWithSet || !property.getType().equals( ValueTypes.PROPERTY_SET );
            if ( valueTypeRestrictionSatisfied && propertyWithSetRestrictionSatisfied )
            {
                visit( property );
            }

            if ( property.getType().equals( ValueTypes.PROPERTY_SET ) )
            {
                if ( property.hasNotNullValue() )
                {
                    traverse( property.getSet() );
                }
            }
        }
    }

    public abstract void visit( final Property property );
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/data/PropertyVisitor.ts
module api.data {

    export class PropertyVisitor {

        private valueType: ValueType;

        public restrictType(value: ValueType): PropertyVisitor {
            this.valueType = value;
            return this;
        }

        public traverse(propertySet: PropertySet) {

            propertySet.forEach((property: Property, index: number) => {

                if (this.valueType == null || this.valueType == property.getType()) {
                    this.visit(property);
                }

                if (property.getType().equals(ValueTypes.DATA)) {
                    if (property.hasNonNullValue()) {
                        this.traverse(property.getPropertySet());
                    }
                }
            });
        }

        public visit(property: Property) {
            throw new Error("Must be implemented by inheritor");
        }
    }
}
